### ISSUE TITLE
How to draw Chinese characters?

### ISSUE THREAD

**Issue by betenner** [crosire/scripthookvdotnet#257]
I want to draw some Chinese characters using _DRAW_TEXT, but all I got are some squares (like missing font).
I used the default font (0: GTA.Font.ChaletLondon) and game language is set to Chinese Traditional.
Is there anything to do with the charset? C# uses Unicode charset as default but GTA uses ANSI. I tried using MarshalAs(UnmanagedType.LPStr) but no use.
Any idea on how to resove this?
Thanks.

**Comment by crosire** [comment]
I doubt GTA supports unicode characters, so there isn't much you can do.

**Comment by betenner** [comment]
GTA may not support unicode, but it DOES support Chinese characters.
The native trainer has been translated into Chinese and it works fine.
The LPCSTR string written in an unicode .cpp file can be displayed in GTA (it has been confirmed).
I just wonder how to do it in C#, maybe you can do something when you convert the C# string and pass it into native function so that it can be displayed correctly? (maybe convert it into ANSI string will work?)

**Comment by kagikn** [comment]
GTA might uses UTF-8. I haven't confirmed it yet, though.

**Comment by betenner** [comment]
I need a sample code in C# with ScriptHookVDotNet that actually display Chinese characters in game successfully (with game language set to Chinese Traditional).

**Comment by GrandTheftMods** [comment]
download notepad++  search your character online copy it into np++  and change the format to ANSI   copy the converted string/char as displayed in np++ into your mod ....

**Comment by GrandTheftMods** [comment]
枪 = æžª       as an example

**Comment by betenner** [comment]
So just use "æžª" in my c# code and it will work?

**Comment by betenner** [comment]
GrandTheftMods: I tried your method, it didn't work. Can you give me a working sample code in c#? Thanks!

**Comment by GrandTheftMods** [comment]
http://www11.pic-upload.de/20.07.15/v1875p2ss793.jpg   its working fine for me (i've german game language)   (äöü are normally displayed as squares....  so i think it should work with chinese chars as well)

**Comment by betenner** [comment]
To GrandTheftMods: I saw your pic, but they are not chinese characters..., I tried your method but didn't work. Can you set your game language to Chinese and try some chinese characters to see whether it works? Thanks.

**Comment by betenner** [comment]
To GrandTheftMods: And what font did you use?

**Comment by kagikn** [comment]
It's Chalet Comprime Cologne 1960, but GTA doesn't contain this font in CJK version(vanilla).

**Comment by betenner** [comment]
I just wonder why the native trainer can show chinese characters correctly?
Just use LPCSTR string with chinese characters such as LPCSTR str = "玩家選項", use _DRAW_TEXT to draw that and save the cpp file in Unicode charset, compile and run. I've tested it and confirmed it.
I want to know how to achieve that in c# with scripthookv.net.

**Comment by crosire** [comment]
LPCSTR is a 8-bit char string format, meaning those characters apparently have to be UTF8 (which is usually what happens if you save the file in a unicode charset). But C# strings are UTF16, that gets you completly different results. Try something like this:
```c#
string str = Encoding.UTF8.GetString(Encoding.UTF8.GetBytes("玩家選項"));
```

**Comment by betenner** [comment]
TO crosire: Thanks, I'll try that later and tell you the result.

**Comment by betenner** [comment]
To crosire: It doesn't work, still display squares. Can you make me a working sample code? Thanks.

**Comment by betenner** [comment]
@crosire Can you add byte[] to InputArgument as string so that I can directly pass the encoded string to GTA? Maybe this will work?

**Comment by crosire** [comment]
Unicode strings are now handled correctly as of ae6780a2e187535594e3dd714541d2380ccf7854. Checked against a bunch of different unicode characters, which did not work before. Your chinese characters still show up as squares for me, for which the only explanation left is that the font simply does not support those, so they are replaced with a placeholder character (the square).

**Comment by betenner** [comment]
@crosire Then why native code can display Chinese characters? (Localized native trainer can show them correctly, use only LPCSTR with UTF-8 encoding...)

**Comment by betenner** [comment]
@crosire PS: The game supports Chinese Traditional language, you can set it in Steam. And the game displays menu and subtitle in Chinese Traditional, so I think this means GTA V does support Chinese characters in a certain way.

**Comment by crosire** [comment]
Just checked with the language set to chinese again (-uilanguage chinese) and now they do show up as expected with the latest commit.

**Comment by crosire** [comment]
Fixed as of ae6780a2e187535594e3dd714541d2380ccf7854.

**Comment by betenner** [comment]
@crosire When will you publish a new release that contain this fix?
